### PR TITLE
add sni fields in FlowInformation

### LINKED ISSUE #14: add sni field

### LINKED ISSUE #16: Access to the entire flow map from the Generator
I need to browse all the map to create the features.

### PR DESCRIPTION
**Pull request by PascaleBorscia**
close #14 #15 #16

### CODE REVIEW

File: src/flow_information.rs [hunk 1]
```diff
@@ -6,6 +6,8 @@ use crate::packet::Packet;
 /// It contains forward and backward packet's.
 #[derive(Debug, Default, Deserialize, Serialize)]
 pub struct FlowInformation {
+    /// SNI field
+    pub sni: String,
```
Comment (tigroo31) [inline src/flow_information.rs]: I would prefer Option<&str>:

Optional to be able to easily handle the missing values
&str to reduce the payload (String gathers a lot of methods...useless here).

File: src/flow_information.rs [hunk 2]
```diff
@@ -26,13 +28,15 @@ mod tests {
     #[test]
     fn test_default() {
         let default = FlowInformation::default();
+        assert!(default.sni.is_empty());
```
Comment (tigroo31) [inline src/flow_information.rs]: If Option<&str>, let's check the field is there empty/ not empty/not there if possible

File: src/generator.rs [hunk 3]
```diff
@@ -104,6 +104,7 @@ mod tests {
         "transport_protocol": 17
       },
       {
+        "sni": "",
```
Comment (tigroo31) [inline src/generator.rs]: If Option, let's check the field isn't serialized as null if None. Let's use the #[serde_with::skip_serializing_none].

**Review by tigroo31 — CHANGES_REQUESTED** [review]
Tests ok. Code reviewed.

File: src/flow_information.rs (hunk 1 above)
Comment (PascaleBorscia) [reply src/flow_information.rs]: you 're right but it's a little more difficult with lifetime parameter. I modify.

File: src/flow_information.rs (hunk 2 above)
Comment (PascaleBorscia) [reply src/flow_information.rs]: ok

File: src/generator.rs (hunk 3 above)
Comment (PascaleBorscia) [reply src/generator.rs]: ok

File: src/flow_information.rs (hunk 1 above)
Comment (PascaleBorscia) [reply src/flow_information.rs]: we keep String. 
the type becomes Option<String>

File: src/generator.rs [hunk 4]
```diff
@@ -53,6 +53,21 @@ impl Generator {
...
         self.flow_map.entry(key)
     }
+
+    /// used for IntoIterator trait
+    pub fn add(&mut self, key: FlowId, value: FlowInformation) {
+        self.flow_map.insert(key, value);
+    }
+}
+
+impl IntoIterator for Generator {
```
Comment (tigroo31) [inline src/generator.rs:63]: the 2 commits are on the same files and the same subject: I'm going to squash tthem merging the PR.

File: src/generator.rs [hunk 5]
```diff
@@ -568,8 +583,108 @@ mod tests {
...
+        assert_eq!(generator.len(), 2);
+
+        let expected_flow_id_1 = FlowId::new(132, "10.216.31.192", "10.15.194.132", 0, 0);
+        let expected_flow_id_2 = FlowId::new(
+            6,
+            "2a01:cb06:a02d:8571:4706:7df1:bd62:5169",
+            "2a00:1450:4007:810::2004",
+            42254,
+            443,
+        );
```
Comment (tigroo31) [inline src/generator.rs:611]: all of that may may factorized with "it_can_write_a_complete_flow_map" into a "create_complete_flow" test method (no code duplication)

File: src/generator.rs [hunk 6]
```diff
@@ -568,8 +583,108 @@ mod tests {
...
+        for (i, (flow_id, _flow_information)) in generator.into_iter().enumerate() {
+            n = i;
+            let result = panic::catch_unwind(|| {
+                assert_eq!(flow_id, expected_flow_id_1);
+            });
+            if result.is_err() {
+                assert_eq!(flow_id, expected_flow_id_2);
+            }
+        }
+        assert_eq!(n + 1, 2);
```
Comment (tigroo31) [inline src/generator.rs]: A few func prog to simplify the code (no counter):

```
assert_eq!(
            generator.into_iter().fold(0, |n, (flow_id, _flow_information)| {
                let result = panic::catch_unwind(|| {
                    assert_eq!(flow_id, expected_flow_id_1);
                });
                if result.is_err() {
                    assert_eq!(flow_id, expected_flow_id_2);
                };
                n + 1
            }),
            2
        );
```

fold is your friend ;).

**Review by tigroo31 — CHANGES_REQUESTED** [review]
Tests ok.
Code reviewed: let's factorize flow creation and use fold into the new unit test.
Please let's squash all the changes into a same commit (or at least do not change many times the same source line through many commit) to keep a clean commit tree.

File: src/generator.rs (hunk 4 above)
Comment (PascaleBorscia) [reply src/generator.rs:63]: 👍

File: src/generator.rs (hunk 6 above)
Comment (PascaleBorscia) [reply src/generator.rs]: 👍

File: src/generator.rs (hunk 5 above)
Comment (PascaleBorscia) [reply src/generator.rs:611]: create create_generator_with_complete_flow called in the two tests

File: src/generator.rs [hunk 7]
```diff
@@ -559,8 +590,37 @@ mod tests {
...
+
+        let expected_flow_id_1 = FlowId::new(132, "10.216.31.192", "10.15.194.132", 0, 0);
+        let expected_flow_id_2 = FlowId::new(
+            6,
+            "2a01:cb06:a02d:8571:4706:7df1:bd62:5169",
+            "2a00:1450:4007:810::2004",
+            42254,
+            443,
+        );
+        let mut n = 0;
```
Comment (tigroo31) [inline src/generator.rs]: Useless. I remove it by myself and accept the PR.

**Review by tigroo31 — APPROVED** [review]
Code reviewed and tested.

**Pull request merged**